### PR TITLE
Improvement with Daybreak detection

### PR DESCRIPTION
**Pull request by shadow2560**
I've changed how Daybreak is detected, personnaly it is in "/switch/daybreak/daybreak.nro" cause I prefer to not place NROs directly in my "/switch" folder. It is effectivly slower than before but I've tried to minimize the impact of the modification and for me it's better cause it doesn't force a specific config for this function to work except having Daybreak correctly set in the homebrews list.
Like I've said in a previous issue I have a Daybreak implementation in one of my homebrew witch could be easily integrated in Sphaira with some smalls modifications, if you want I can try to do it? Or do you plan to make your own implementation of this yourself?

### CODE REVIEW

**Comment by ITotalJustice** [comment]
You don't need to scan for nros, they're already scanned in the homebrew menu. There's a reference to that vector in the filebrowser menu to use, which I use for file assoc.

I won't reimplement daybreak in sphaira, sorry. Daybreak works with no bugs, and is bundled with atmosphere.

As for changing the daybreak path, I'm not sure why you do this 😄 won't it break every atmosphere update? I'd rather not accept this because in the event daybreak "breaks", updating atmosphere will fix this by simply overwriting the old version. Where as if you have 2 locations for it, there's a chance to select the broken version when launching.

Basically, I'd rather keep the standardized paths in place :)

**Comment by shadow2560** [comment]
I'm not placing any nro in "/switch" folder directly because the logic to place the nro in it folder simplify the read of the files, for example if I want to know if my homebrew is correctly stared or if I want to verify some configs I know exactly where to search. Lots of homebrews work with some config files in their folders placed in the "/switch" folder, with my logic I don't need to search where the nro is, where the config is, if a homebrew has a folder or not, everything is configured the same way. Maybe these preferences are linked to my visual impairement, navigating one element after an other even with a good knowlage with the keyboard navigation is more simple when having a minimum of logics set, you you can read X lines of elements in one time and quicly review what you want.
That's why I've implemented this, if you prefer I can force the path "/switch/daybreak.nro" to be prefered and if not found or is not Daybreak try to find an other path?

**Comment by HenryBaby** [comment]
> I'm not placing any nro in "/switch" folder directly because the logic to place the nro in it folder simplify the read of the files, for example if I want to know if my homebrew is correctly stared or if I want to verify some configs I know exactly where to search. Lots of homebrews work with some config files in their folders placed in the "/switch" folder, with my logic I don't need to search where the nro is, where the config is, if a homebrew has a folder or not, everything is configured the same way. Maybe these preferences are linked to my visual impairement, navigating one element after an other even with a good knowlage with the keyboard navigation is more simple when having a minimum of logics set, you you can read X lines of elements in one time and quicly review what you want. That's why I've implemented this, if you prefer I can force the path "/switch/daybreak.nro" to be prefered and if not found or is not Daybreak try to find an other path?

I do the exact same thing. It's just more tidy to me. It's a bummer to update stuff, but it beats having a cluttered switch folder.

**Comment by ITotalJustice** [comment]
> I'm not placing any nro in "/switch" folder directly because the logic to place the nro in it folder simplify the read of the files, for example if I want to know if my homebrew is correctly stared or if I want to verify some configs I know exactly where to search. Lots of homebrews work with some config files in their folders placed in the "/switch" folder, with my logic I don't need to search where the nro is, where the config is, if a homebrew has a folder or not, everything is configured the same way. Maybe these preferences are linked to my visual impairement, navigating one element after an other even with a good knowlage with the keyboard navigation is more simple when having a minimum of logics set, you you can read X lines of elements in one time and quicly review what you want. That's why I've implemented this, if you prefer I can force the path "/switch/daybreak.nro" to be prefered and if not found or is not Daybreak try to find an other path?

After some consideration, I've decided I'll add support for having daybreak be in a custom path. Like you suggested, it will check the default path first, otherwise it will scan the nro vector for daybreak. I'd like to submit the fix myself if that's okay with you? I don't mean to disregard your pr, it's just that you don't have to scan folders/files for daybreak as all nros are already scanned, and thus it won't have any impact on performance :)

**Comment by shadow2560** [comment]
No problem, you know better than me how to optimize your code, I'll be hapy to look how you will make this and learn more about the homebrew. Thanks for your concideration about this and all of my issues/pull requests, I realy apreciate.

**Comment by shadow2560** [comment]
Just for info I've done a modification witch use the vector used by the file explorer and the function search_homebrew() doesn't launch a rescan, it just search the homebrew in the list passed:
https://github.com/ITotalJustice/sphaira/pull/56/commits/5f22ffef0b8aae0f809b22d682da9cfba0abe954

File: sphaira/source/nro.cpp [hunk 1]
```diff
@@ -309,4 +309,19 @@ auto nro_normalise_path(const std::string& p) -> std::string {
...
 
+auto search_homebrew(const std::vector<NroEntry> m_entries, const char* name, const fs::FsPath path) -> NroEntry {
+    for (auto& p : m_entries) {
+        if (path == "") {
+            if (!std::strcmp(p.nacp.lang[0].name, name)) {
+                return p;
+            }
+        } else {
+            if (!std::strcmp(p.path.s, path.s) && !std::strcmp(p.nacp.lang[0].name, name)) {
+                return p;
```
Comment (ITotalJustice) [inline sphaira/source/nro.cpp:320]: the switch filesystem is case-insensitive, meaning a.nro and A.nro are the same.
you should use strcasecmp whenever comparing files names

File: sphaira/include/nro.hpp [hunk 2]
```diff
@@ -76,4 +76,6 @@ auto nro_add_arg_file(std::string arg) -> std::string;
 // strips sdmc:
 auto nro_normalise_path(const std::string& p) -> std::string;
 
+auto search_homebrew(const std::vector<NroEntry> m_entries, const char* name, const fs::FsPath path = "") -> NroEntry;
```
Comment (ITotalJustice) [inline sphaira/include/nro.hpp:79]: this passes the vector by value, meaning it re-creates the entire vector in memory.

also, the return type should use std::optional to signify that the value returned may not be valid.

File: sphaira/source/ui/menus/filebrowser.cpp [hunk 3]
```diff
@@ -105,7 +105,7 @@ constexpr RomDatabaseEntry PATHS[]{
     { "wonderswancolor", "Bandai - WonderSwan Color"},
 };
 
-constexpr fs::FsPath DAYBREAK_PATH{"/switch/daybreak.nro"};
```
Comment (ITotalJustice) [inline sphaira/source/ui/menus/filebrowser.cpp:108]: you should keep the path. when trying to find daybreak, first check the path and then find the nro as fallback. no need to make this global :)

File: sphaira/source/ui/menus/filebrowser.cpp [hunk 4]
```diff
@@ -252,12 +252,21 @@ auto GetRomIcon(ProgressBox* pbox, std::string filename, std::string extension,
 }
 
 // returns 0 if true
-auto CheckIfUpdateFolder(const fs::FsPath& path, std::span<FileEntry> entries) -> Result {
+auto CheckIfUpdateFolder(const fs::FsPath& path, std::span<FileEntry> entries, const std::vector<NroEntry> nros) -> Result {
     fs::FsNativeSd fs;
     R_TRY(fs.GetFsOpenResult());
 
     // check that we have daybreak installed
-    R_UNLESS(fs.FileExists(DAYBREAK_PATH), FsError_FileNotFound);
```
Comment (ITotalJustice) [inline sphaira/source/ui/menus/filebrowser.cpp:260]: see above comments

**Comment by ITotalJustice** [comment]
i made a branch here https://github.com/ITotalJustice/sphaira/commit/5c9b8a8cfa38598a4fd4662692464dc35fb91adf. is this okay for you?

**Comment by shadow2560** [comment]
I've thinked to also verify the author of the homebrew to be more sure that it's the good one.
Your modifications are far better than mine, I've so much to learn, I'm a little frustrated to have finaly done it so badly. So sure, it's OK for me and thanks to have reviewed my code and pointed the problems.

**Comment by ITotalJustice** [comment]
Your code wasn't bad at all. Just needed a few changes :)
